### PR TITLE
Add graceful shutdown to prometheus server

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Fixes prometheus server not stopping if there are open connections

### CODE REVIEW

File: substrate/utils/prometheus/src/lib.rs [hunk 1]
```diff
@@ -111,7 +111,11 @@ async fn init_prometheus_with_listener(
 		}
 	});
 
-	let server = Server::builder(listener).serve(service);
+	// Shutdown server when "signal" is dropped
+	let (_signal, on_exit) = exit_future::signal();
```
Comment (bkchr) [inline substrate/utils/prometheus/src/lib.rs:115]:
```suggestion
	let (signal, on_exit) = tokio::sync::oneshot::channel::<()>();
```

File: substrate/utils/prometheus/src/lib.rs [hunk 2]
```diff
@@ -111,7 +111,11 @@ async fn init_prometheus_with_listener(
...
+	let server = Server::builder(listener).serve(service).with_graceful_shutdown(async {
+		on_exit.await;
+	});
 
 	let result = server.await.map_err(Into::into);
```
Comment (bkchr) [inline substrate/utils/prometheus/src/lib.rs:120]:
```suggestion
	let result = server.await.map_err(Into::into);
	
	let _ = signal.send(());
```

File: substrate/utils/prometheus/Cargo.toml [hunk 3]
```diff
@@ -13,6 +13,7 @@ readme = "README.md"
 targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
+exit-future = "0.2.0"
```
Comment (bkchr) [inline substrate/utils/prometheus/Cargo.toml:16]:
```suggestion
```

**Review by bkchr — APPROVED** [review]

**Review by davxy — APPROVED** [review]

File: substrate/utils/prometheus/src/lib.rs (hunk 1 above)
Comment (davxy) [reply substrate/utils/prometheus/src/lib.rs:115]: I also prefer the one shot chan solution

File: substrate/utils/prometheus/src/lib.rs (hunk 2 above)
Comment (tmpolaczyk) [reply substrate/utils/prometheus/src/lib.rs:119]: Thanks, I have implemented it this way, and I have verified it also works.

**Pull request merged**